src/btree/bt_page.c

```diff
@@ -102,10 +102,23 @@ __wt_page_in_func(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags
 			return (WT_RESTART);
 		case WT_REF_MEM:
 			/*
-			 * The page is in memory: get a hazard pointer, update
-			 * the page's LRU and return.  The expected reason we
-			 * can't get a hazard pointer is because the page is
-			 * being evicted; yield and try again.
+			 * The page is in memory.
+			 *
+			 * Check if we need an autocommit transaction.
+			 */
+			WT_RET(__wt_txn_autocommit_check(session));
```

Comment:
We shouldn't start the transaction for real until after resolving forced evictions.  Otherwise, we're guaranteed to have a running transaction in the session that is evicting a large page (or waiting for some other thread to evict it).

Maybe a `goto`?